### PR TITLE
[wip] Avoid adding redundant type members in incremental mode

### LINKED ISSUE #1164: (From Fuzzer) (Incremental resolve) core/types/types.cc:632 enforced condition this->klass.data(ctx)->typeMembers().size() == this->targs.size() || (this->klass == Symbols::Array() && (this->targs.size() == 1)) || (this->klass == Symbols::Hash() && (this->targs.size() == 3)) || this->klass._id >= Symbols::Proc0()._id && this->klass._id <= Symbols::last_proc()._id has failed: <C <U FullChild>>
#### Input

[→ View on sorbet.run](https://sorbet.run/#class%20FullChild%20V%3Dtype_member%0Asig%7Bparams%28f%3AFullChild%29%7D%0AV%3Dy%0Aend)

```ruby
class FullChild V=type_member
sig{params(f:FullChild)}
V=y
end
```

#### Observed output

With `--stress-incremental-resolver`:
```
/Users/jvilk/stripe/sorbet/newest_fuzzers/crash-0d3fee2a0b306fc707f10f8617a78dc6f3c50be1.done:3: Redefining constant `V` https://srb.help/4012
     3 |V=y
        ^^^
    /Users/jvilk/stripe/sorbet/newest_fuzzers/crash-0d3fee2a0b306fc707f10f8617a78dc6f3c50be1.done:1: Previous definition
     1 |class FullChild V=type_member
                        ^^^^^^^^^^^^^
Exception::raise(): core/types/types.cc:632 enforced condition this->klass.data(ctx)->typeMembers().size() == this->targs.size() || (this->klass == Symbols::Array() && (this->targs.size() == 1)) || (this->klass == Symbols::Hash() && (this->targs.size() == 3)) || this->klass._id >= Symbols::Proc0()._id && this->klass._id <= Symbols::last_proc()._id has failed: <C <U FullChild>>

Backtrace:
sorbet::core::AppliedType::_sanityCheck(sorbet::core::Context) (in libcore.so) (types.cc:628)
sorbet::core::Type::sanityCheck(sorbet::core::Context) (in libinfer.so) (Types.h:238)
sorbet::resolver::TypeSyntax::getResultTypeAndBind(sorbet::core::MutableContext, sorbet::ast::Expression&, sorbet::resolver::ParsedSig const&, bool, bool, sorbet::core::SymbolRef) (in libresolver.so) (type_syntax.cc:809)
sorbet::resolver::TypeSyntax::parseSig(sorbet::core::MutableContext, sorbet::ast::Send*, sorbet::resolver::ParsedSig const*, bool, sorbet::core::SymbolRef) (in libresolver.so) (type_syntax.cc:223)
sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk::processLeftoverSigs(sorbet::core::MutableContext, absl::InlinedVector<sorbet::ast::Send*, 1ul, std::__1::allocator<sorbet::ast::Send*> >&) (in libresolver.so) (resolver.cc:964)
sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk::processClassBody(sorbet::core::MutableContext, std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >&) (in libresolver.so) (resolver.cc:981)
sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk::postTransformClassDef(sorbet::core::MutableContext, std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >) (in libresolver.so) (resolver.cc:1333)
sorbet::ast::PostPonePostTransform_ClassDef<sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk, sorbet::core::MutableContext, true>::call(sorbet::core::MutableContext, std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk&) (in libresolver.so) (treemap.h:221)
sorbet::ast::TreeMapper<sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk, sorbet::core::MutableContext>::mapClassDef(std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::core::MutableContext) (in libresolver.so) (treemap.h:285)
sorbet::ast::TreeMapper<sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk, sorbet::core::MutableContext>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::MutableContext) (in libresolver.so) (treemap.h:672)
sorbet::ast::TreeMapper<sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk, sorbet::core::MutableContext>::mapInsSeq(std::__1::unique_ptr<sorbet::ast::InsSeq, std::__1::default_delete<sorbet::ast::InsSeq> >, sorbet::core::MutableContext) (in libresolver.so) (treemap.h:601)
sorbet::ast::TreeMapper<sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk, sorbet::core::MutableContext>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::MutableContext) (in libresolver.so) (treemap.h:668)
std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> > sorbet::ast::TreeMap::apply<sorbet::core::MutableContext, sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk>(sorbet::core::MutableContext, sorbet::resolver::(anonymous namespace)::ResolveSignaturesWalk&, std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >) (in libresolver.so) (treemap.h:712)
sorbet::resolver::Resolver::resolveSigs(sorbet::core::MutableContext, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >) (in libresolver.so) (resolver.cc:1568)
sorbet::resolver::Resolver::runTreePasses(sorbet::core::MutableContext, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >) (in libresolver.so) (resolver.cc:1597)
sorbet::realmain::pipeline::incrementalResolve(sorbet::core::GlobalState&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&) (in libpipeline.so) (pipeline.cc:321)
sorbet::realmain::pipeline::resolve(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&, bool) (in libpipeline.so) (pipeline.cc:874)
sorbet::realmain::realmain(int, char**) (in librealmain.so) (realmain.cc:429)
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1

Exception::raise(): main/pipeline/pipeline.cc:875 enforced condition reresolved.size() == 1 has failed: (no message provided)

Backtrace:
sorbet::Exception::enforce_handler(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) (in librealmain.so) (Exception.h:47)
sorbet::realmain::pipeline::resolve(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&, bool) (in libpipeline.so) (pipeline.cc:875)
sorbet::realmain::realmain(int, char**) (in librealmain.so) (realmain.cc:429)
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1


/Users/jvilk/stripe/sorbet/newest_fuzzers/crash-0d3fee2a0b306fc707f10f8617a78dc6f3c50be1.done:2: Malformed type declaration. Generic class without type arguments `FullChild` https://srb.help/5004
     2 |sig{params(f:FullChild)}
                     ^^^^^^^^^

/Users/jvilk/stripe/sorbet/newest_fuzzers/crash-0d3fee2a0b306fc707f10f8617a78dc6f3c50be1.done:2: Malformed `sig`. No method def following it https://srb.help/5003
     2 |sig{params(f:FullChild)}
        ^^^^^^^^^^^^^^^^^^^^^^^^

/Users/jvilk/stripe/sorbet/newest_fuzzers/crash-0d3fee2a0b306fc707f10f8617a78dc6f3c50be1.done:64: Failed to process tree (backtrace is above) https://srb.help/1001
    64 |class FullChild V=type_member
    65 |sig{params(f:FullChild)}
    66 |V=y
    67 |end

???: Exception resolving (backtrace is above) https://srb.help/1001

???: Exception resolving (backtrace is above) https://srb.help/1001
Errors: 6
```

#### Expected behavior

Exception does not occur.

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
Fixes #1164: when we had a situation in which if we redefined type members, we would later on add "new" type members when we re-ran the namer (in incremental mode) because the existing type member array contained only the redefined constant, which in turn was given a unique name: because the uniquified name `T$1` is not `==` to `T`, we would believe that we were seeing a "new" type variable and re-add it.

This fixes it with a kind of broad hammer: when looking to see if we've already defined a type member, we look not at the unique name, but rather at the original `NameRef` if it exists. This does mean looking through the `data` for the name, and my feeling is that this will be rare enough (only happening on type member definitions and only looking through a small set of names) that it's probably alright, but I still have some minor FUD about whether this is the best way of solving this problem.

### Test plan

See included automated tests. Added the fuzzed test case.

### CODE REVIEW

File: namer/namer.cc [hunk 1]
```diff
@@ -703,7 +703,9 @@ class NameInserter {
         }
 
         auto members = onSymbol.data(ctx)->typeMembers();
-        auto it = absl::c_find_if(members, [&](auto mem) { return mem.data(ctx)->name == typeName->cnst; });
+        auto it = absl::c_find_if(members, [&](auto mem) {
```
Comment (DarkDimius) [inline namer/namer.cc]: I don't understand what this does. Could you explain?
Why do you only compare `originalName`?

File: core/Names.cc [hunk 2]
```diff
@@ -243,6 +243,14 @@ string NameRef::show(const GlobalState &gs) const {
     return data(gs)->show(gs);
 }
 
+NameRef NameRef::originalName(const GlobalState &gs) const {
```
Comment (DarkDimius) [inline core/Names.cc]: Please don't introduce globally accessible methods without a clear semantic usecase for them outside of a signle module.

**Comment by aisamanra** [comment]
This is both on reflection and based on your feedback absolutely the wrong solution, but I should explain the problem step-by-step a bit more:

```
class X
  # at this point, our typeMembers() array is empty
  T = T.type_member  # after this, our typeMembers() array contains [T]
  T = x  # when we look at this symbol, we realize it already exists, so we use
         # `mangleRenameSymbol` to rename it to the unique name T$1. This also
         # means that our typeMembers() array contains [T$1].
end

# now, because we're using the incremental resolver, we walk this code again
# with the namer

class X
  # remember, our typeMembers() array contains [T$1]
  T = T.type_member  # we look to see whether T is present in typeMembers().
                     # it's not, only T$1 is, so we add it: now it contains [T$1, T]
                     # (This is why this bad fix works: even though the thing we've
                     # stored is T$1, we know that it was "originally" T, so we can
                     # determine that it's a duplicate now.)
  T = x  # This does not modify the typeMembers() array any more, so it still has
         # the same two it had before
end
```

**Comment by DarkDimius** [comment]
[From IRL on Thursday, so that I  don't forget]
crasher without incremental mode
```
# typed: true

class Enumerable
  S = type_member
end

[1].flatten
```

**Comment by aisamanra** [comment]
This is being addressed by a different PR entirely.

**Pull request closed without merging**